File: 프로그래머스/2/131701. 연속 부분 수열 합의 개수/연속 부분 수열 합의 개수.py

def solution(elements):
    # 길이가 1~len(elements)까지인거 모두 구해서 카운팅하기
    # 결과값을 저장해서 이미 존재하면 카운팅에서 제외
    n = len(elements)
    elements.extend(elements)
    #print(elements)
    
    sums = set([])
    cnt = 0
    for length in range(1, n+1): # 길이
        for start in range(n): # 시작점
            temp = sum(elements[start:start+length])
            if not temp in sums:
                cnt += 1
                sums.add(temp)
    return cnt

# 시작점을 먼저 잡으면 이전 합에 더하여 더 효율적으로 구할 수 있음
        
    

File: 프로그래머스/2/131701. 연속 부분 수열 합의 개수/연속 부분 수열 합의 개수.java
import java.util.*;
class Solution {
    public int solution(int[] elements) {
        /*
        길이가 1 ~ n까지인 부분수열 순차적으로 구해서 저장
        n번째 길이면 n-1 번째 것 이전에 구한 합에 더하기
        */
        Set<Integer> score = new HashSet<>();
        int n = elements.length;
        int[] sum = new int[n];
        for(int i=0; i<n; i++){ // 더할 인덱스
            for(int j=0; j<n; j++){ // 시작 인덱스
                sum[j] += elements[(j+i)%n];
                if(!score.contains(sum[j])){
                    score.add(sum[j]);
                }
            }
        }
        return score.size();
    }
}